src/main/java/graphql/execution/ExecutionStrategy.java

```diff
@@ -120,25 +122,29 @@ protected ExecutionResult completeValue(ExecutionContext executionContext, Graph
         return executionContext.getQueryStrategy().execute(executionContext, resolvedType, result, subFields);
     }
 
-    private ExecutionResult completeValueForList(ExecutionContext executionContext, GraphQLList fieldType, List<Field> fields, Object result) {
+    private ExecutionResult completeValueForList(ExecutionContext executionContext, GraphQLList fieldType, List<Field> fields, Object result, Map<String, Object> argumentValues) {
         if (result.getClass().isArray()) {
             result = Arrays.asList((Object[]) result);
         }
 
         //noinspection unchecked
-        return completeValueForList(executionContext, fieldType, fields, (Iterable<Object>) result);
+        return completeValueForList(executionContext, fieldType, fields, (Iterable<Object>) result, argumentValues);
     }
 
-    protected GraphQLObjectType resolveType(GraphQLInterfaceType graphQLInterfaceType, Object value) {
-        GraphQLObjectType result = graphQLInterfaceType.getTypeResolver().getType(value);
+    protected GraphQLObjectType resolveType(Field field, GraphQLInterfaceType graphQLInterfaceType,
+                                            Object value, Map<String, Object> argumentValues, GraphQLSchema schema) {
+        TypeResolutionEnvironment env = new TypeResolutionEnvironment(value, argumentValues, field, graphQLInterfaceType, schema);
+        GraphQLObjectType result = graphQLInterfaceType.getTypeResolver().getType(env);
         if (result == null) {
             throw new GraphQLException("could not determine type");
         }
         return result;
     }
 
-    protected GraphQLObjectType resolveType(GraphQLUnionType graphQLUnionType, Object value) {
```

Comment:
Not sure I got your 2nd suggestion... But anything that allows easier non-breaking modifications sounds like a good idea.